knowledge-base/adr/0032-monorepo-tooling.md

```diff
@@ -0,0 +1,131 @@
+- By default, Turborepo will run scripts for downstream dependencies (e.g. test packages), but not for upstream dependencies (e.g. odata-common).
+  `--no-deps` ignores downstream and `--include-dependencies` adds upstream dependencies.
+  - Think of this as "the dependencies didn't change so we don't need to run them, but we need to make sure to not break our downstream dependants"
+- Turborepo does not handle releasing in any form (yet). A separate library like **Changesets**, Lerna, or Beachball is needed.
+- Turborepo will detect and fail if there are cycles in the package dependency graph.
+  - This is currently the case and needs to be worked around by using peerDeps in our test packages. It would be better if we fix this in the future.
+- Significantly less downloads (164k weekly) than `nx` (1.4M weekly)
+
+Benefits:
+
+- Performance of scripts is increased dramatically through parallelization and caching (PR checks turbo 6min vs lerna 9min, PR tests turbo 2-3min vs lerna ~5min)
+- `turbo.json` makes it possible to understand dependencies of scripts.
+  - There are still improvements possible to the changes in the PR like generating test services before testing (and using the cache to avoid slowing it down).
+- It is easier to run scripts for all packages, simplifying the generate and readme scripts.
+- Tests will not fail because some other action (`yarn compile` or `yarn generate`) was not run before executing tests
+- Good documentation is available and there is a helpful community.
+- There is active development backed by a committed third party (Vercel).
+- [Powerful filters](https://turborepo.org/docs/features/filtering) allow us to run commands in all relevant packages.
+
+### Changesets
+
+Caveats:
+
+- Need to maintain changeset files (instead of `CHANGELOG.md`).
+- We need special permission to use the bot and need to verify if we can use the action. We could work around this, but this is extra effort.
+- Our current versioning strategy means we have fixed versions for all packages.
+  Changesets would support linked versions, which would avoid meaningless version increases for our packages.
+  We should reconsider our versioning policy given this new opportunity.
+- It maintains one `CHANGELOG.md` per package, but allows custom changelog formatting.
+- When creating a changeset, affected packages need to be selected.
+  This adds some extra effort.
+  - There is a "changed packages" selection, that makes it easy to select all packages that were touched since the last changeset.
+
+<img width="560" alt="changeset console output" src="https://user-images.githubusercontent.com/26565533/164425056-581e4cd6-6863-4d8c-b78f-c13ceabf3b23.png">
+
+Benefits:
+
+- Documentation is good but contains many TODOs.
+- CLI is amazingly simple:
+  - Create a changeset with `yarn changeset`.
+  - Apply changeset with `yarn changeset version`.
+  - Create release with `yarn changeset publish` or `yarn changeset publish --tag beta`.
```

Comment:
Serious Question: How we plan to do the intermediate period? Let's say we merge this PR tonight, then there are already unpublished changes on the main branch which are not modeled in the per package changeset. However, our release script relies now on the `changeset` - will this cause problem?

Reply:
Or do we simply do a bit handy work for the next release on the generated changelog file and after this the lib will work fine because there are no untracked changes?

Reply:
Good point. I will migrate the changes we have so far into the changesets format. We may need a check before the next release that nothing was added without a changeset, but that should be low effort.